extensions/scarb-snforge-test-collector/src/compilation/test_collector.rs

```diff
@@ -0,0 +1,248 @@
+use config::{forge_try_extract_test_config, FuzzerConfig, SingleTestConfig};
+pub use config::{ForkConfig, RawForkConfig};
+use plugin::snforge_test_plugin_suite;
+use sierra_casm_generator::SierraCasmGenerator;
+
+mod config;
+mod plugin;
+mod sierra_casm_generator;
+
+fn find_all_tests(
+    db: &dyn SemanticGroup,
+    main_crate: CrateId,
+) -> Vec<(FreeFunctionId, SingleTestConfig)> {
+    let mut tests = vec![];
+    let modules = db.crate_modules(main_crate);
+    for module_id in modules.iter() {
+        let Ok(module_items) = db.module_items(*module_id) else {
+            continue;
+        };
+        tests.extend(module_items.iter().filter_map(|item| {
+            let ModuleItemId::FreeFunction(func_id) = item else {
+                return None;
+            };
+            let Ok(attrs) = db.function_with_body_attributes(FunctionWithBodyId::Free(*func_id))
+            else {
+                return None;
+            };
+            Some((
+                *func_id,
+                forge_try_extract_test_config(db.upcast(), &attrs).unwrap()?,
+            ))
+        }));
+    }
+    tests
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub struct TestCase<T: ForkConfig> {
+    pub name: String,
+    pub available_gas: Option<usize>,
+    pub ignored: bool,
+    pub expected_result: ExpectedTestResult,
+    pub fork_config: Option<T>,
+    pub fuzzer_config: Option<FuzzerConfig>,
+}
+
+type TestCaseRaw = TestCase<RawForkConfig>;
+
+pub fn collect_tests(
+    crate_name: &str,
+    crate_root: &Path,
+    lib_content: &str,
+    compilation_unit: &CompilationUnit,
+) -> Result<(Program, Vec<TestCaseRaw>)> {
+    let crate_roots: OrderedHashMap<SmolStr, PathBuf> = compilation_unit
+        .dependencies()
+        .iter()
+        .cloned()
+        .map(|source_root| (source_root.name.into(), source_root.path))
+        .collect();
+
+    let project_config = ProjectConfig {
+        base_path: crate_root.into(),
+        corelib: Some(Directory::Real(compilation_unit.corelib_path()?)),
+        content: ProjectConfigContent {
+            crate_roots,
+            crates_config: compilation_unit.crates_config_for_compilation_unit(),
+        },
+    };
+
+    // code taken from crates/cairo-lang-test-runner/src/lib.rs
+    let db = &mut {
+        let mut b = RootDatabase::builder();
+        b.with_cfg(CfgSet::from_iter([Cfg::name("test")]));
+        b.with_plugin_suite(snforge_test_plugin_suite());
+        b.with_plugin_suite(test_plugin_suite());
+        b.with_plugin_suite(starknet_plugin_suite());
+        b.with_project_config(project_config);
+        b.build()?
+    };
+
+    let main_crate_id =
+        insert_lib_entrypoint_content_into_db(db, crate_name, crate_root, lib_content);
+
+    if DiagnosticsReporter::stderr().check(db) {
+        return Err(anyhow!(
+            "Failed to compile test artifact, for detailed information go through the logs above"
+        ));
+    }
```

Comment:
FYI https://docs.rs/anyhow/latest/anyhow/macro.ensure.html

Reply:
I think it would make it less readable